Added a working AddIngredientView

gojamesgo85 opened the pull request:
Added an "AddIngredientView" that would add to the working ingredients array in the CREATEACOCKTAILVIEWMODEL. 

ADDINGREDIENTVIEW populates a list of CocktailComponents to choose from that are a list of buttons. Once selected, the button will add the component to an @Published var in the CREATEACOCKTAILVIEWMODEL. 

When the user hits the "Add to Ingredients" button, the information in the amount field, the measurement unit menu, and the ingredient name field are then converted into a CocktailIngredient and then added to the array. 

This solves for ingredients that exist in the app. I have yet to find a solution to new ingredients.

MatthewHunt84 reviewed (changes requested):
Tons of work here, it's impressive.

Couple of bits of feedback:

I'm not sure I agree with adding the year to the cocktail name. We already have the year in most cases stored in the author. So because that data already exists in the model, adding the date to the name makes things a bit messier.

I would say revert those changes, and then instead add a new CocktailCollection case called "originals":
![image](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/96077863/1b1c01cf-19a9-451e-b196-f89ce157aba0)
Then you can just add that collection to those cocktails instead, since it's already a part of the model.
![image](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/96077863/9990d70b-c8b8-447e-adae-05282aaf0c9f)

Then, after we pull the cocktails from swift data we can do something like
_if cocktail.collection = .originals && cocktail.author.year =! nil { name = cocktail.cocktailName + " (\(cocktail.author.year) recipe)") }_ or whatever when we want to display it that way.

I think Prep is a cool addition, but lets make those properties static, so we don't have to initiate a bunch of new viewModels everytime we add one. Also that can just be a class, it's not a viewmodel since it doesn't have any presentation logic:

![image](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/96077863/03f651a9-89d2-4d6e-ad20-8b977744a77c)

Observation changes are great, glad that's working it's much nicer syntax!

Finally, for the addIngredientView here are my proposed changes:

![image](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/96077863/920a7250-eba0-4db7-89e8-7981a3b810f0)

Optionally, I would also make that textfield take numbers only and make the search view show nothing until something is typed (instead of the whole list like we have it now)

gojamesgo85 commented:
I'm just working on making the last couple of changes requested. I'm getting hung up on the search view show nothing until something is typed.

all of the "matchedCurrentSearch" bools are set to true because of the init on CocktailComponent
![Screenshot 2024-02-29 at 1 16 23 AM](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/95845784/354e2258-85bf-4572-afe2-709f17e5b74a)


When the view initially loads, I can set all of the values to false by calling viewModel.matchAllPhysicalCocktailComponents(),but that doesn't get done until after the view is drawn so the list is still shown.

![Screenshot 2024-02-29 at 1 20 13 AM](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/95845784/53f7f21e-a09e-4421-85cf-b30a9b57774b)

So then I thought to set the focus on the text field when the view loads, focing a view update because of the onchange modifier on SearchBarView. So I created an @FocusState var.
![Screenshot 2024-02-29 at 1 24 58 AM](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/95845784/8f6b1a7f-d6bd-4a79-b3ab-c835b5b17e4b)

then I added that modifier to the SearchBarView
![Screenshot 2024-02-29 at 1 25 49 AM](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/95845784/ecedab18-6127-4583-b141-ede2c93d3e2e)

then I set keyboardFocused = true on .task.

![Screenshot 2024-02-29 at 1 26 54 AM](https://github.com/MatthewHunt84/MetaCocktailsSwiftData/assets/95845784/d636cec4-af55-4360-9a0a-14a9507a9092)

That works, kinda. 

What happens is the view loads showing the full list, then a quarter of a second in, the view disappears when the text field becomes focused because the .task doesn't get called until the rest of the view above it gets drawn. 

Do you have some ideas or can you give me some direction with this one?

gojamesgo85 commented:
Everything is changed that you've requested. I still can't figure out the results business.

gojamesgo85 commented:
Added more cocktails from Milk and Honey.

I temporarily removed the @Attribute(.unique) from the Cocktail model because it kept crashing after I added the two Apple Jack cocktails. I don't understand this one. I checked to make sure that the names and everything else were unique, and they are. But for some reason, it still crashes. 

I have it commented out on the model so you can fuck with it. I looked at everything. I'm at a loss for that one.

MatthewHunt84 commented on MetaCocktailsSwiftData/Views/CreateACocktailView/CreateACocktailViewModel.swift:
```diff
@@ -51,18 +51,18 @@ import Observation
              selectedMeasurementUnit = MeasurementUnit.drops
              
+         } else {
+             selectedMeasurementUnit = MeasurementUnit.fluidOunces
```
Lets make this a switch statement instead

MatthewHunt84 reviewed (changes requested):
Nice work man, there's so much here god damn!


I fixed that attribute bug and pushed up the change. When we add new cocktails, we just have to add them to the preload file - that's all that was. It wasn't caused by the applejack cocktail - that was just the first cocktail from the list that it couldn't find in the model because it wasn't added in the preload step.

Requested a couple of minor changes, then merge it in!
